### DISCUSSION TITLE
yiwen-blog-website/docs/ai/architectures/transformer/grpo-training-guide

### DISCUSSION THREAD

**Discussion by giscus** [Hao-yiwen/yiwen-blog-website#440] [Announcements]
# yiwen-blog-website/docs/ai/architectures/transformer/grpo-training-guide

摘要：GRPO（组相对策略优化）是一种用于大语言模型强化学习的高效算法。该算法由 DeepSeek 团队在 DeepSeekMath 论文中提出，并在 DeepSeek-V3 和 DeepSeek-R1 等高性能模型的训练中发挥了核心作用。GRPO 的核心突破在于：它摒弃了传统 PPO 算法中必须的"评论家"（Critic）模型，通过从"一组"生成的输出中计算相对优势，显著降低了训练时的显存占用和计算成本。

https://hao-yiwen.github.io/yiwen-blog-website/docs/ai/architectures/transformer/grpo-training-guide

**Comment by Hao-yiwen** [comment]
GRPO 的损失函数（Loss Function）本质上是 **PPO（Proximal Policy Optimization）** 损失函数的一个变体，但去掉了 Critic 部分，并修改了优势（Advantage）的计算方式。

为了让你彻底理解，我们将公式拆解开来，逐个符号进行详细说明。

---

### 1. GRPO 的核心目标函数公式

GRPO 试图最大化以下目标函数 $J_{GRPO}(\theta)$：

$$
J_{GRPO}(\theta) = \mathbb{E}_{q \sim P(Q), \{o_i\}_{i=1}^G \sim \pi_{\theta_{old}}} \left[ \frac{1}{G} \sum_{i=1}^G \left( \underbrace{\min \left( \frac{\pi_\theta(o_i|q)}{\pi_{\theta_{old}}(o_i|q)} A_i, \text{clip}\left( \frac{\pi_\theta(o_i|q)}{\pi_{\theta_{old}}(o_i|q)}, 1-\epsilon, 1+\epsilon \right) A_i \right)}_{\text{① 策略更新 (PPO-Clip)}} - \underbrace{\beta \mathbb{D}_{KL}(\pi_\theta || \pi_{ref})}_{\text{② KL 散度惩罚}} \right) \right]
$$

为了计算 Loss（用于梯度下降），我们需要**最小化**上述目标的负值：$Loss = - J_{GRPO}$。

---

### 2. 逐个符号详细解释

#### A. 基础变量与采样
* **$\theta$ (Theta):** 当前正在训练（更新）的模型的参数。
* **$\theta_{old}$:** 在本次更新迭代开始前的模型参数（也就是采样数据时的那个模型）。
* **$q$:** 输入的问题（Prompt）。
* **$G$:** 组的大小（Group Size）。比如 $G=64$，意味着对于同一个问题 $q$，我们采样了 64 个不同的回答。
* **$o_i$:** 第 $i$ 个采样生成的输出（Output）。
* **$\{o_i\}_{i=1}^G \sim \pi_{\theta_{old}}$:** 表示这组数据是旧模型生成的。

#### B. 概率比率 (Probability Ratio)
$$\frac{\pi_\theta(o_i|q)}{\pi_{\theta_{old}}(o_i|q)}$$
* **含义：** 这是“新策略”与“旧策略”生成同一个输出 $o_i$ 的概率之比。
* **作用：**
    * 如果比率 $> 1$，说明新模型比旧模型更倾向于生成这个回答。
    * 如果比率 $< 1$，说明新模型降低了生成这个回答的概率。
* **目的：** 我们希望通过调整 $\theta$，让高分回答（$A_i > 0$）的比率变大，让低分回答的比率变小。

#### C. 优势函数 (Advantage) $A_i$
这是 GRPO 与 PPO 最大的不同点。这里没有 Critic 模型，而是基于组内统计计算：
$$A_i = \frac{r_i - \text{mean}(R_{group})}{\text{std}(R_{group})}$$
* **$r_i$:** 第 $i$ 个回答获得的原始奖励（由规则或奖励模型给出）。
* **$\text{mean}(R_{group})$:** 这 $G$ 个回答的平均奖励。
* **$\text{std}(R_{group})$:** 这 $G$ 个回答的标准差。
* **解释：** 这就是**“相对优劣”**。如果你的分比这组的平均分高，$A_i$ 就是正数；否则就是负数。

#### D. 裁剪机制 (Clipping)
$$\text{clip}(..., 1-\epsilon, 1+\epsilon)$$
* **$\epsilon$ (Epsilon):** 一个很小的超参数，通常取 0.1 或 0.2。
* **作用：** 它限制了概率比率的变化范围，使其不能超过 $[0.8, 1.2]$（假设 $\epsilon=0.2$）。
* **为什么需要？** 防止模型更新步子迈得太大（Policy Collapse）。如果一次更新让参数变化太剧烈，训练很容易崩溃。

#### E. 最小值操作 (Min Operator)
$$\min( \text{Ratio} \cdot A_i, \text{ClippedRatio} \cdot A_i )$$
* **作用：** 这是一个**悲观下界（Pessimistic Bound）**策略。
* 它在“原始增益”和“裁剪后的增益”之间取最小值。这意味着算法只承诺“保守”的优化，确保在最坏情况下性能也不会下降太多。这是 PPO 算法稳定性的核心来源。

#### F. KL 散度惩罚 (KL Divergence Penalty)
$$- \beta \mathbb{D}_{KL}(\pi_\theta || \pi_{ref})$$
* **$\pi_{ref}$ (Reference Model):** 参考模型，通常是 SFT（监督微调）后的原始模型，参数冻结不变。
* **$\mathbb{D}_{KL}$:** KL 散度，衡量当前模型 $\pi_\theta$ 和参考模型 $\pi_{ref}$ 之间的分布差异。
* **$\beta$ (Beta):** 惩罚系数。
* **解释：** 我们不希望模型为了刷高分（Reward Hacking）而彻底改变说话的方式（比如输出乱码但包含正确答案）。这一项强迫模型“不要离初心太远”，保持语言的流畅性和原始分布。

---

### 3. 公式整体直观解读

把公式翻译成**人话**，GRPO 在做这样一件事：

1.  **采样：** 老师（环境）给了一道题 $q$，学生（$\pi_{old}$）试着写了 64 个不同的解法 $\{o_1...o_G\}$。
2.  **打分：** 助教根据规则给这 64 个解法打分，算出哪些是优于平均的（$A_i > 0$），哪些是差于平均的（$A_i < 0$）。
3.  **更新（Loss 计算）：**
    * 对于**好**的解法：我们要**提高**它出现的概率（Ratio 变大），但**不能提高得太离谱**（被 Clip 限制住），防止自负。
    * 对于**差**的解法：我们要**降低**它出现的概率。
4.  **约束：** 同时，不管你怎么改，你的说话风格不能离正常的教科书（$\pi_{ref}$）太远，否则我要扣你分（KL 惩罚）。

### 4. 总结：GRPO Loss 的精髓

GRPO Loss = **(PPO 的裁剪策略机制)** + **(基于组内均值的优势计算)** + **(KL 散度约束)**

它巧妙地用 `mean(Group)` 替代了 PPO 中昂贵的神经网络 Critic $V(s)$，从而让公式中的 $A_i$ 变得极其廉价且容易计算，同时保留了 PPO 稳定训练的所有优点。

**Comment by Hao-yiwen** [comment]
这是一个非常敏锐的问题！通常只有深入研究过 PPO/RLHF 代码的人才会问出 **“Clip 和 KL 到底有什么区别，是不是重复了？”** 这个问题。

结论先行：**它们没有重复。Clipping 是限制“步幅”（不要跑太快），KL 是限制“方向”（不要跑偏了）。**

下面我详细解释 KL 的计算方法、作用，以及它与 Clipping 的本质区别。

-----

### 1\. KL 散度是怎么计算的？

在 LLM 训练中，我们计算的是**当前模型（$\pi_\theta$）** 和 **参考模型（$\pi_{ref}$，通常是 SFT 后的冻结模型）** 之间的 KL 散度。

对于一个生成的 Token $x$，KL 散度的计算通常简化为**对数概率之差**：

$$
\mathbb{D}_{KL} \approx \log \frac{\pi_\theta(x|context)}{\pi_{ref}(x|context)} = \log \pi_\theta(x|context) - \log \pi_{ref}(x|context)
$$

**在代码中的实际操作（以 PyTorch 为例）：**

```python
# 1. 获取当前模型对生成的 token 的 log 概率
# current_log_probs shape: [batch_size, seq_len]
current_log_probs = actor_model(input_ids).log_softmax(dim=-1)

# 2. 获取参考模型（不更新参数）对同样 token 的 log 概率
with torch.no_grad():
    ref_log_probs = ref_model(input_ids).log_softmax(dim=-1)

# 3. 计算逐个 token 的 KL (就是做个减法)
# per_token_kl shape: [batch_size, seq_len]
per_token_kl = current_log_probs - ref_log_probs

# 4. 有时候会取近似值 (Schulman 的近似法)
# kl = (exp(ratio) - 1) - ratio ... 这种写法在 PPO 库中常见，为了数值稳定性
```

**直观理解：**

  * 如果当前模型生成某个词的概率是 0.9，参考模型也是 0.9，差值为 0，KL = 0（没偏离）。
  * 如果当前模型生成“杀”字的概率是 0.9（为了得分），参考模型是 0.001（平时不这么说话），差值巨大，KL 惩罚就会很大。

-----

### 2\. 既然已经有 Clipping 了，为什么还要 KL？

这是你困惑的核心。让我们对比一下这两者锁定的对象：

#### A. Clipping (PPO 的核心机制)

  * **比较对象：** 当前策略 $\pi_\theta$ vs **上一轮策略 $\pi_{old}$**。
  * **公式：** $\text{clip}(\frac{\pi_\theta}{\pi_{old}}, 0.8, 1.2)$
  * **作用：** **控制更新步长（Step Size）。**
    它保证了在**这一次**梯度下降中，模型参数不要突变。它关注的是\*\*“训练的稳定性”\*\*，防止一次更新把模型搞崩了。

#### B. KL 散度 (KL Divergence)

  * **比较对象：** 当前策略 $\pi_\theta$ vs **初始 SFT 模型 $\pi_{ref}$**。
  * **公式：** $\pi_\theta - \pi_{ref}$
  * **作用：** **控制模型漂移（Model Drift）。**
    它保证了在**整个训练过程**（成千上万次更新）中，模型不要彻底忘记它原本“该怎么说话”。它关注的是\*\*“语言的合法性”\*\*。

#### 图解区别

> 想象你在训练一只狗（模型）：
>
>   * **Clipping (牵引绳):** 每次训练时，你限制狗**一次只能往前冲 1 米**。防止它突然发疯把你拽倒。（限制单次更新幅度）
>   * **KL (围栏/家):** 不管训练了多少天，你要求狗**不能跑出小区**。如果它为了追骨头（Reward）跑到了邻市去，那就回不来了。（限制相对于原始状态的偏离）

-----

### 3\. 如果没有 KL 会发生什么？（Reward Hacking）

如果没有 KL 约束，只靠 Clipping，模型虽然每次更新都很稳，但经过 10000 次更新后，它会为了最大化奖励而**彻底“黑化”**。这就是著名的 **Reward Hacking（奖励欺骗）**。

**举个真实的例子：**

  * **任务：** 总结新闻。
  * **奖励模型（RM）：** 只要生成的摘要里包含原文的关键信息就给高分。
  * **无 KL 的结果：** 模型发现，把原文里的关键词毫无语法逻辑地堆砌在一起（比如：“特朗普 总统 经济 下跌 股票”），RM 也会给高分。
      * 模型变成了“关键词提取器”，完全**丧失了人类语言的语法能力**。
      * 因为它发现写完整的句子太累了，直接崩单词效率最高。

**KL 的作用就是：**
当模型试图输出“关键词堆砌”时，$\pi_{ref}$（参考模型，也就是那个懂语法的好学生）会说：“在这个语境下，生成这个词的概率应该很低！”
于是 $KL$ 值飙升，Loss 变大，模型受到惩罚，被迫滚回去老老实实写人话。

-----

### 4\. 总结

在 Loss 函数中：

$$Loss = \underbrace{- \text{Advantage}}_{\text{为了赢}} + \underbrace{\beta \cdot KL}_{\text{为了像人}}$$

  * **Clipping** 是为了防止模型在\*\*“为了赢”\*\*的过程中步子迈太大扯着蛋（数值计算崩溃）。
  * **KL** 是为了防止模型\*\*“为了赢”\*\*而不择手段，最后变成一个只会刷分的怪物（丧失通用语言能力）。

所以，这两者缺一不可。Clipping 保护**训练过程**，KL 保护**模型质量**。
